cpp/src/plasma/external_store_worker.cc

```diff
@@ -0,0 +1,202 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include <arrow/util/memory.h>
+#include <utility>
+
+#include "plasma/external_store_worker.h"
+#include "plasma/plasma.h"
+
+namespace plasma {
+
+ExternalStoreWorker::ExternalStoreWorker(std::shared_ptr<ExternalStore> external_store,
+                                         const std::string& external_store_endpoint,
+                                         const std::string& plasma_store_socket,
+                                         size_t parallelism)
+    : valid_(external_store != nullptr),
+      plasma_store_socket_(plasma_store_socket),
+      plasma_clients_(parallelism, nullptr),
+      parallelism_(parallelism),
+      sync_handle_(nullptr),
+      async_handles_(parallelism, nullptr),
+      terminate_(false),
+      stopped_(false),
+      num_writes_(0),
+      num_bytes_written_(0),
+      num_reads_not_found_(0),
+      num_reads_(0),
+      num_bytes_read_(0) {
+  if (valid_) {
+    ARROW_CHECK_OK(external_store->Connect(external_store_endpoint, &sync_handle_));
```

Comment:
We shouldn't just fail here without giving the user a chance to recover. The status can't be returned in the constructor, so we probably need a Connect or Initialize method here.